csharp/src/Ice/TcpEndpoint.cs

```diff
@@ -2,284 +2,216 @@
-                        }
-
-                        _compress = true;
-
-                        return true;
+                        throw new FormatException($"invalid timeout value `{argument}' in endpoint `{endpointString}'",
+                             ex);
                     }
-
-                default:
+                    if (Timeout < 1)
                     {
-                        return false;
+                        throw new FormatException($"invalid timeout value `{argument}' in endpoint `{endpointString}'");
                     }
+                }
+                options.Remove("-t");
+            }
+            else
+            {
+                Timeout = Instance.DefaultTimeout;
             }
-        }
 
-        protected override IConnector CreateConnector(EndPoint addr, INetworkProxy? proxy) =>
-            new TcpConnector(Instance, addr, proxy, SourceAddr, _timeout, ConnectionId_);
+            if (options.TryGetValue("-z", out argument))
+            {
+                if (argument != null)
+                {
+                    throw new FormatException(
+                        $"unexpected argument `{argument}' provided for -z option in `{endpointString}'");
+                }
+                HasCompressionFlag = true;
+                options.Remove("-z");
+            }
+        }
 
-        protected override IPEndpoint CreateEndpoint(string? host, int port, string connectionId) =>
-            new TcpEndpoint(Instance, host!, port, SourceAddr, _timeout, connectionId, _compress);
+        private protected override IConnector CreateConnector(EndPoint addr, INetworkProxy? proxy) =>
+            new TcpConnector(Instance, addr, proxy, SourceAddress, Timeout, ConnectionId);
 
-        private int _timeout;
-        private bool _compress;
+        private protected override IPEndpoint CreateEndpoint(string host, int port, string connectionId) =>
+            new TcpEndpoint(Instance, host, port, SourceAddress, Timeout, connectionId, HasCompressionFlag);
     }
 
     internal sealed class TcpEndpointFactory : IEndpointFactory
     {
-        internal TcpEndpointFactory(TransportInstance instance) => _instance = instance;
+        private TransportInstance _instance;
 
         public void Initialize()
         {
         }
-
-        public EndpointType Type() => _instance!.Type;
-
-        public string Transport() => _instance!.Transport;
-
-        public Endpoint Create(List<string> args, bool oaEndpoint)
+        public void Destroy()
         {
-            IPEndpoint endpt = new TcpEndpoint(_instance!);
-            endpt.InitWithOptions(args, oaEndpoint);
-            return endpt;
         }
 
-        public Endpoint Read(Ice.InputStream s) => new TcpEndpoint(_instance!, s);
+        public EndpointType Type() => _instance.Type;
+        public string Transport() => _instance.Transport;
```

Comment:
Shouldn't these be properties and should we rename Transport to Name instead?

Reply:
Yes, I left these for a future PR.